### ISSUE TITLE
Java 12 functional sanity failure TestFlushReflectionCache_0

### ISSUE THREAD

**Issue by pshipton** [eclipse/openj9#4658]
I ran a Java 12 functional sanity test using the changes from #3980 and #4531.

https://ci.eclipse.org/openj9/job/Test-sanity.functional-JDK12-linux_390-64_cmprssptrs/3
TestFlushReflectionCache_0
```
===============================================
Running test TestFlushReflectionCache_0 ...
===============================================
TestFlushReflectionCache_0 Start Time: Thu Feb  7 16:02:43 2019 Epoch Time (ms): 1549573363431
variation: NoOptions
JVM_OPTIONS: -Xcompressedrefs 
[AgentMain] [INFO] AgentMain.premain() is initilized
[IncludeExcludeTestAnnotationTransformer] [INFO] exclude file is /home/jenkins/jenkins-agent/workspace/Test-sanity.functional-JDK12-linux_390-64_cmprssptrs/openjdk-tests/TestConfig/scripts/testKitGen/../../../../jvmtest/TestConfig/resources/excludes/latest_exclude_12.txt
...
... TestNG 6.14.2 by Cédric Beust (cedric@beust.com)
...

[Cmvc196982] [INFO] Cmvc196982 is setting up... 
FAILED CONFIGURATION: @BeforeMethod setUp
java.lang.NoSuchFieldException: annotations
	at java.base/java.lang.Class.getDeclaredField(Class.java:901)
	at org.openj9.test.javaagenttest.Cmvc196982.setUp(Cmvc196982.java:85)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:59)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:455)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:222)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:520)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:716)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:988)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:648)
	at org.testng.TestRunner.run(TestRunner.java:505)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:455)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:450)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:415)
	at org.testng.SuiteRunner.run(SuiteRunner.java:364)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:84)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1208)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1137)
	at org.testng.TestNG.runSuites(TestNG.java:1049)
	at org.testng.TestNG.run(TestNG.java:1017)
	at org.testng.TestNG.privateMain(TestNG.java:1354)
	at org.testng.TestNG.main(TestNG.java:1323)

SKIPPED: testAddingAnnotationsFlushesReflectCache
java.lang.NoSuchFieldException: annotations
	at java.base/java.lang.Class.getDeclaredField(Class.java:901)
	at org.openj9.test.javaagenttest.Cmvc196982.setUp(Cmvc196982.java:85)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:59)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:455)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:222)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:520)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:716)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:988)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:648)
	at org.testng.TestRunner.run(TestRunner.java:505)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:455)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:450)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:415)
	at org.testng.SuiteRunner.run(SuiteRunner.java:364)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:84)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1208)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1137)
	at org.testng.TestNG.runSuites(TestNG.java:1049)
	at org.testng.TestNG.run(TestNG.java:1017)
	at org.testng.TestNG.privateMain(TestNG.java:1354)
	at org.testng.TestNG.main(TestNG.java:1323)
```

**Comment by pshipton** [comment]
@theresa-m @DanHeidinga @JasonFengJ9

**Comment by JasonFengJ9** [comment]
The `NoSuchFieldException: annotations` was at `methodAnnotationsField = Method.class.getDeclaredField("annotations");`.

This is due to `JDK12` tights the access to more sensitive fields. 
https://github.com/ibmruntimes/openj9-openjdk-jdk12/blob/caa80184d0d8e75fd982c43af78cc2a26259d8b5/src/java.base/share/classes/jdk/internal/reflect/Reflection.java#L41-L63

Here is `JDK11` and earlier versions:
https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/afa710e1295efc46b0284835a7a1643f9cebc702/src/java.base/share/classes/jdk/internal/reflect/Reflection.java#L42-L58

So the test need to be updated to avoid these fields/methods.

**Comment by pshipton** [comment]
@JasonFengJ9 can we exclude this test, at least temporarily, to get green test results for jdk12.

**Comment by JasonFengJ9** [comment]
@pshipton the test in question depends on reflection of field `annotations` to verify the test results:
https://github.com/eclipse/openj9/blob/6ae06288ecc080788394732958aae3748321109f/test/functional/JavaAgentTest/src/org/openj9/test/javaagenttest/Cmvc196982.java#L137-L140
I didn't see a diagnosis switch to workaround this restriction however the field in question can be retrieved via package access method [java.lang.reflect.Method.getAnnotationBytes()](
https://github.com/ibmruntimes/openj9-openjdk-jdk12/blob/e969f57ff8031485b00ccf2ed10f87a627c7bd68/src/java.base/share/classes/java/lang/reflect/Method.java#L211-L214).
It is interesting to note that this method hasn't been filtered by [jdk/internal/reflect/Reflection.methodFilterMap](https://github.com/ibmruntimes/openj9-openjdk-jdk12/blob/e969f57ff8031485b00ccf2ed10f87a627c7bd68/src/java.base/share/classes/jdk/internal/reflect/Reflection.java#L62). I guess this could be changed at any time before `JDK12` is finalized.

Will exclude this test for `JDK12+`.

**Comment by pshipton** [comment]
If the test can be changed to use getAnnotationBytes(), that is preferred. It will be a couple of days before all the failures are addressed, so excluding it quickly doesn't really provide any benefit over fixing it.

**Comment by JasonFengJ9** [comment]
This test `org.openj9.test.javaagenttest.Cmvc196982` requires reflection access to following `class/field`:
https://github.com/eclipse/openj9/blob/8095d8078c473a3815e31a16c69b9590d6b062d0/test/functional/JavaAgentTest/src/org/openj9/test/javaagenttest/Cmvc196982.java#L85-L86

`Method.class.getDeclaredField("annotations")` can be workaround via `Method.java:byte[] getAnnotationBytes()`, however there is no such method for `Field.class.getDeclaredField("annotations")`, hence this test has to be excluded for `JDK12+`.

**Comment by JasonFengJ9** [comment]
Attempted to reflect the fields in question via package access methods or skip the field without such method, and at last hit same issue as https://github.com/eclipse/openj9/issues/4663 due to following code snippet:
https://github.com/eclipse/openj9/blob/4577abf9392aecd6574ca4a5a84c6761894123d2/test/functional/JavaAgentTest/src/org/openj9/test/javaagenttest/Cmvc196982.java#L210

These tests need to be copied to `Java12andUp` (being added at https://github.com/eclipse/openj9/pull/4105) and update to `jdk.internal.org.objectweb.asm.Opcodes.ASM7` which is not available for `JDK8`.

**Comment by pshipton** [comment]
> These tests need to be copied to `Java12andUp` (being added at #4105) and update to `jdk.internal.org.objectweb.asm.Opcodes.ASM7`

@JasonFengJ9 are you working on this?

**Comment by JasonFengJ9** [comment]
@pshipton will pick it up today.

**Comment by JasonFengJ9** [comment]
This will be test only change, can we move it out `Release 0.13.0 (Java 12)` to favour other higher priority items?

**Comment by pshipton** [comment]
@JasonFengJ9 if it can' be completed by the 19th then we'll move it out. Until then, agreed, please focus on higher priority items.

**Comment by pshipton** [comment]
Seems clear this won't be addressed in the 0.13 release, moving to 0.14
